### PR TITLE
generate voter id #220

### LINKED ISSUE #220: Gerar um número Válido e Aleatório de Título Eleitoral
**Seu pedido de recurso está relacionado a um problema? Por favor, descreva.**
Desejamos adicionar um novo utilitário à nossa biblioteca que permita aos usuários gerar números de Títulos Eleitorais válidos e aleatórios, onde eles possam escolher o estado da emissão.

**Descreva a solução que você gostaria**
- Chamar um novo utilitário passando opções para o estado de emissão e receber um número de Título Eleitoral válido e aleatório.
```
- O número de Inscrição do Título Eleitoral, em geral, possui doze dígitos.
- Os oito primeiros dígitos correspondem ao número sequencial.
- O nono e o décimo dígitos correspondem à Unidade Federativa (UF) à qual o eleitor pertence. Se o título não for expedido no Brasil, considera-se que a sigla da UF é 28 (ZZ).
- Os dois últimos dígitos correspondem aos Dígitos Verificadores (DV’s).
```

**Descreva alternativas que você considerou**
- O utilitário deve aceitar as seguintes opções:
- Estado: correspondente ao Estado de emissão desejado. O padrão é 28 (estrangeiro) se não for especificado.

- Criar um novo utilitário `generate_titulo_eleitoral`:
```python
def generate_titulo_eleitoral(estado):
 ...
```

Exemplo de uso
```python
from brutils import generate_titulo_eleitoral

>>> generate_titulo_eleitoral()
"630227662877"  # Estado não identificado = Estrangeiro (28)
>>> generate_titulo_eleitoral(estado=SP)
"307801530159"  # Estado SP = 01
>>> generate_titulo_eleitoral(estado=BA)
"438286340507"  # Estado BA =  05
>>> generate_titulo_eleitoral(estado=MT)
"220130701813"  # Estado MT =  18
```

**Contexto adicional**
-https://pt.wikipedia.org/wiki/T%C3%ADtulo_de_eleitor#:~:text=C%C3%A1lculo%20do%20d%C3%ADgito%20verificador,-Para%20realizar%20o&text=Se%20o%20t%C3%ADtulo%20eleitoral%20n%C3%A3o,d%C3%ADgitos%20correspondem%20ao%20n%C3%BAmero%20sequencial.
- http://clubes.obmep.org.br/blog/a-matematica-nos-documentos-titulo-de-eleitor/

### PR DESCRIPTION
**Pull request by patricia-salles**
## Descrição
Objetivo dessa PR é incluir a função que será responsável por gerar um título de eleitor válido e aleatório.

## Mudanças Propostas
Inclusão da função generate_voter_id() e ajuste na função _calculate_vd2() para incluir a verificação se o resto da divisão for igual a 10.

## Checklist de Revisão

- [x] Eu li o [Contributing.md](https://github.com/brazilian-utils/brutils-python/blob/main/CONTRIBUTING.md)
- [x] Os testes foram adicionados ou atualizados para refletir as mudanças (se aplicável).
- [x] Foi adicionada uma entrada no changelog / Meu PR não necessita de uma nova entrada no changelog.
- [x] A [documentação](https://github.com/brazilian-utils/brutils-python/blob/main/README.md) em português foi atualizada ou criada, se necessário.
- [x] Se feita a documentação, a atualização do [arquivo em inglês](https://github.com/brazilian-utils/brutils-python/blob/main/README_EN.md). 
- [x] Eu documentei as minhas mudanças no código, adicionando docstrings e comentários. [Instruções](https://github.com/brazilian-utils/brutils-python/blob/main/CONTRIBUTING.md#8-fa%C3%A7a-as-suas-altera%C3%A7%C3%B5es)
- [x] O código segue as diretrizes de estilo e padrões de codificação do projeto.
- [x] Todos os testes passam. [Instruções](https://github.com/brazilian-utils/brutils-python/blob/main/CONTRIBUTING.md#testes)
- [x] O Pull Request foi testado localmente. [Instruções](https://github.com/brazilian-utils/brutils-python/blob/main/CONTRIBUTING.md#7-execute-o-brutils-localmente)
- [x] Não há conflitos de mesclagem.

## Issue Relacionada

Closes #220

### CODE REVIEW

**Comment by patricia-salles** [comment]
Olá @antoniamaia, adicionei a nova PR da issue para gerar um número de título de eleitor :)

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Utilitário `format_pis` [#224](https://github.com/brazilian-utils/brutils-python/pull/224)
 - Utilitário `format_legal_process` [#210](https://github.com/brazilian-utils/brutils-python/pull/210)
 - Utilitário `generate_license_plate` [#241](https://github.com/brazilian-utils/brutils-python/pull/241)
+<<<<<<< HEAD
```
Comment (camilamaia) [inline CHANGELOG.md]:
```suggestion
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -43,6 +44,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Utilitário `remove_symbols_phone` [#188](https://github.com/brazilian-utils/brutils-python/pull/188)
 - Utilitário `remove_symbols_pis` [#236](https://github.com/brazilian-utils/brutils-python/pull/236)
 - Utilitário `remove_symbols_legal_process` [#209](https://github.com/brazilian-utils/brutils-python/pull/209)
+=======
+>>>>>>> 6e22034d1f9918376b786c463e9974c6d2f4da2b
```
Comment (camilamaia) [inline CHANGELOG.md]:
```suggestion
```

File: README.md [hunk 3]
```diff
@@ -79,13 +79,23 @@ False
   - [remove_symbols_pis](#remove_symbols_pis)
   - [generate_pis](#generate_pis)
 - [Processo Jurídico](#processo-jurídico)
+<<<<<<< HEAD
```
Comment (camilamaia) [inline README.md]:
```suggestion
```

File: README.md [hunk 4]
```diff
@@ -79,13 +79,23 @@ False
...
   - [is_valid_voter_id](#is_valid_voter_id)
+=======
+  - [format_processo_juridico](#format_processo_juridico)
+  - [remove\_symbols\_processo\_juridico](#remove_symbols_processo_juridico)
+  - [generate_processo_juridico](#generate_processo_juridico)
+  - [is_valid_processo_juridico](#is_valid_processo_juridico)
+- [Título ELeitoral](#título-eleitoral)
+  - [generate_titulo_eleitoral](#generate_titulo_eleitoral)
+
+>>>>>>> 6e22034d1f9918376b786c463e9974c6d2f4da2b
```
Comment (camilamaia) [inline README.md]:
```suggestion
  - [generate_voter_id](#generate_voter_id)
```

File: README.md [hunk 5]
```diff
@@ -897,6 +907,40 @@ False
 True
 ```
 
+### generate_voter_id
+
+Gerar uma string de dígitos de Título de Eleitor válida aleatória a partir de um estado brasileiro informado.
```
Comment (camilamaia) [inline README.md]:
```suggestion
Gera uma string de dígitos de Título de Eleitor válida aleatória a partir de um estado brasileiro informado.
```

File: README.md [hunk 6]
```diff
@@ -897,6 +907,40 @@ False
...
+Gera um número de título eleitoral válido aleatório dado uma unidade federativa. Por padrão o título é emitido fora do Brasil, caso a unidade federativa não seja inforamda.
+
+```python
+>>> from brutils import generate_titulo_eleitoral
+>>> generate_titulo_eleitoral(state='MG')
+'744770710280'
+```
+
+
+>>>>>>> 6e22034d1f9918376b786c463e9974c6d2f4da2b
```
Comment (camilamaia) [inline README.md]:
```suggestion
```

File: README_EN.md [hunk 7]
```diff
@@ -77,13 +77,22 @@ False
   - [remove_symbols_pis](#remove_symbols_pis)
   - [generate_pis](#generate_pis)
 - [Legal Process](#legal-process)
+<<<<<<< HEAD
```
Comment (camilamaia) [inline README_EN.md]:
```suggestion
```

File: README_EN.md [hunk 8]
```diff
@@ -77,13 +77,22 @@ False
...
 - [Voter ID](#voter-id)
   - [is_valid_voter_id](#is_valid_voter_id)
+=======
+  - [format_processo_juridico](#format_processo_juridico)
+  - [remove\_symbols\_processo\_juridico](#remove_symbols_processo_juridico)
+  - [generate_processo_juridico](#generate_processo_juridico)
+  - [is_valid_processo_juridico](#is_valid_processo_juridico)
+- [Título ELeitoral](#título-eleitoral)
+  - [generate_titulo_eleitoral](#generate_titulo_eleitoral)
+>>>>>>> 6e22034d1f9918376b786c463e9974c6d2f4da2b
```
Comment (camilamaia) [inline README_EN.md]:
```suggestion
  - [generate_voter_id](#generate_voter_id)
```

File: README_EN.md [hunk 9]
```diff
@@ -877,6 +886,7 @@ Example:
 >>> generate_legal_process(year=2024, orgao=4)
 "33158248820244017105"
 ```
+<<<<<<< HEAD
```
Comment (camilamaia) [inline README_EN.md]:
```suggestion
```

File: README_EN.md [hunk 10]
```diff
@@ -888,6 +898,9 @@ It does not verify if the voter id actually exists.
 References:
 - https://pt.wikipedia.org/wiki/T%C3%ADtulo_de_eleitor,
 - http://clubes.obmep.org.br/blog/a-matematica-nos-documentos-titulo-de-eleitor/
+=======
+## is_valid_processo_juridico
+>>>>>>> 6e22034d1f9918376b786c463e9974c6d2f4da2b
```
Comment (camilamaia) [inline README_EN.md]:
```suggestion
```

File: README_EN.md [hunk 11]
```diff
@@ -905,6 +918,41 @@ False
...
+
+Generate a valid random voter registration number given a federative unit. The argument has default value (foreigner) if the federative unit is not informed.
+
+```python
+>>> from brutils import generate_titulo_eleitoral
+>>> generate_titulo_eleitoral(state='MG')
+'744770710280'
+```
+
+
```
Comment (camilamaia) [inline README_EN.md]:
```suggestion
```

File: brutils/voter_id.py [hunk 12]
```diff
@@ -204,4 +207,59 @@ def _calculate_vd2(federative_union, vd1):  # type: (str, int) -> str
     if federative_union in ["01", "02"] and rest == 0:
         vd2 = 1
 
+    # edge case: rest == 10, declare vd1 as zero
```
Comment (camilamaia) [inline brutils/voter_id.py]:
```suggestion
    # edge case: rest == 10, declare vd2 as zero
```

File: brutils/voter_id.py [hunk 13]
```diff
@@ -204,4 +207,59 @@ def _calculate_vd2(federative_union, vd1):  # type: (str, int) -> str
...
+
     return vd2
+
+
+def generate(state="ZZ") -> str:
+    """
+    Generates a random valid Brazilian voter registration.
+
+    Args:
+        Federative unit[str]
```
Comment (camilamaia) [inline brutils/voter_id.py]:
```suggestion
        federative_union(str): federative union for the voter id that will be generated. The default value "ZZ" is used for voter IDs issued to foreigners.
```

File: brutils/voter_id.py [hunk 14]
```diff
@@ -204,4 +207,59 @@ def _calculate_vd2(federative_union, vd1):  # type: (str, int) -> str
...
+
+    Returns:
+        value[str]: Brazilian voter registration.
```
Comment (camilamaia) [inline brutils/voter_id.py]:
```suggestion
        str: A randomly generated valid voter ID for the given federative union
```

File: brutils/voter_id.py [hunk 15]
```diff
@@ -204,4 +207,59 @@ def _calculate_vd2(federative_union, vd1):  # type: (str, int) -> str
...
+    }
+
+    state = state.upper()
+    if state in (UFs):
+        sequential_number = str(randint(0, 99999999)).zfill(8)
+        federative_union = UFs[state]
+        if _is_federative_union_valid(federative_union):
+            vd1 = _calculate_vd1(sequential_number, federative_union)
+            vd2 = _calculate_vd2(federative_union, vd1)
+            return f"{sequential_number}{federative_union}{vd1}{vd2}"
```
Comment (camilamaia) [inline brutils/voter_id.py]:
```suggestion
    federative_union = federative_union.upper()
    if federative_union in (UFs):
        sequential_number = str(randint(0, 99999999)).zfill(8)
        uf_number = UFs[federative_union]
        if _is_uf_number_valid(uf_number):
            vd1 = _calculate_vd1(sequential_number, uf_number)
            vd2 = _calculate_vd2(uf_number, vd1)
            return f"{sequential_number}{uf_number}{vd1}{vd2}"
```

File: tests/test_voter_id.py [hunk 16]
```diff
@@ -79,14 +80,35 @@ def test_calculate_vd1(self):
...
         # declare dv2 as 1 instead
         self.assertIs(_calculate_vd2("01", 4), 1)
         # edge case: if UF is "02" (for MG) and rest == 0
         # declare dv2 as 1 instead
         self.assertIs(_calculate_vd2("02", 8), 1)
 
+    def test_generate_voter_id(self):
+        # test if is_valid a voter id from MG
+        voter_id = generate(state="MG")
+
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
```

File: tests/test_voter_id.py [hunk 17]
```diff
@@ -79,14 +80,35 @@ def test_calculate_vd1(self):
...
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if is_valid a voter id from AC
+        voter_id = generate(state="AC")
+
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
```

File: tests/test_voter_id.py [hunk 18]
```diff
@@ -79,14 +80,35 @@ def test_calculate_vd1(self):
...
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if is_valid a voter id from foreigner
+        voter_id = generate()
+
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
```

File: tests/test_voter_id.py [hunk 19]
```diff
@@ -79,14 +80,35 @@ def test_calculate_vd1(self):
...
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if UF is not valid
+        voter_id = generate(state="XX")
+
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
```

File: tests/test_voter_id.py [hunk 20]
```diff
@@ -79,14 +80,35 @@ def test_calculate_vd1(self):
         self.assertIs(_calculate_vd1("73146415", "03"), 0)
 
     def test_calculate_vd2(self):
-        self.assertIs(_calculate_vd2("03", 7), 10)
+        self.assertIs(_calculate_vd2("03", 7), 0)
```
Comment (camilamaia) [inline tests/test_voter_id.py:85]:
```suggestion
        self.assertIs(_calculate_vd2("02", 7), 2)
        # edge case: if rest == 10, declare vd2 as zero
        self.assertIs(_calculate_vd2("03", 7), 0)
```

File: tests/test_voter_id.py [hunk 21]
```diff
@@ -79,14 +80,33 @@ def test_calculate_vd1(self):
...
         # edge case: if UF is "01" (for SP) and rest == 0
         # declare dv2 as 1 instead
         self.assertIs(_calculate_vd2("01", 4), 1)
         # edge case: if UF is "02" (for MG) and rest == 0
         # declare dv2 as 1 instead
         self.assertIs(_calculate_vd2("02", 8), 1)
 
+    def test_generate_voter_id(self):
+        # test if is_valid a voter id from MG
+        voter_id = generate(state="MG")
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
        voter_id = generate(federative_union ="MG")
```

File: tests/test_voter_id.py [hunk 22]
```diff
@@ -79,14 +80,33 @@ def test_calculate_vd1(self):
...
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if is_valid a voter id from AC
+        voter_id = generate(state="AC")
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
        voter_id = generate(federative_union ="AC")
```

File: tests/test_voter_id.py [hunk 23]
```diff
@@ -79,14 +80,33 @@ def test_calculate_vd1(self):
...
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if is_valid a voter id from foreigner
+        voter_id = generate()
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if UF is not valid
+        voter_id = generate(state="XX")
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
        voter_id = generate(federative_union ="XX")
```

File: brutils/voter_id.py [hunk 24]
```diff
@@ -204,4 +207,59 @@ def _calculate_vd2(federative_union, vd1):  # type: (str, int) -> str
...
         vd2 = 1
 
+    # edge case: rest == 10, declare vd2 as zero
+    if rest == 10:
+        vd2 = 0
+
     return vd2
+
+
+def generate(state="ZZ") -> str:
```
Comment (camilamaia) [inline brutils/voter_id.py]:
```suggestion
def generate(federative_union ="ZZ") -> str:
```

File: brutils/voter_id.py [hunk 25]
```diff
@@ -204,4 +207,59 @@ def _calculate_vd2(federative_union, vd1):  # type: (str, int) -> str
...
+        "RR": "26",
+        "TO": "27",
+        "ZZ": "28",
+    }
+
+    federative_union = federative_union.upper()
+    if federative_union in (UFs):
+        sequential_number = str(randint(0, 99999999)).zfill(8)
+        uf_number = UFs[federative_union]
+        if _is_uf_number_valid(uf_number):
```
Comment (camilamaia) [inline brutils/voter_id.py]:
```suggestion
        if _is_federative_union_valid(uf_number):
```

File: brutils/voter_id.py [hunk 26]
```diff
@@ -204,4 +207,59 @@ def _calculate_vd2(federative_union, vd1):  # type: (str, int) -> str
...
         vd2 = 1
 
+    # edge case: rest == 10, declare vd2 as zero
+    if rest == 10:
+        vd2 = 0
+
     return vd2
+
+
+def generate(federative_union ="ZZ") -> str:
```
Comment (camilamaia) [inline brutils/voter_id.py]:
```suggestion
def generate(federative_union="ZZ") -> str:
```

File: tests/test_voter_id.py [hunk 27]
```diff
@@ -79,14 +80,33 @@ def test_calculate_vd1(self):
...
         # edge case: if UF is "01" (for SP) and rest == 0
         # declare dv2 as 1 instead
         self.assertIs(_calculate_vd2("01", 4), 1)
         # edge case: if UF is "02" (for MG) and rest == 0
         # declare dv2 as 1 instead
         self.assertIs(_calculate_vd2("02", 8), 1)
 
+    def test_generate_voter_id(self):
+        # test if is_valid a voter id from MG
+        voter_id = generate(federative_union ="MG")
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
        voter_id = generate(federative_union="MG")
```

File: tests/test_voter_id.py [hunk 28]
```diff
@@ -79,14 +80,33 @@ def test_calculate_vd1(self):
...
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if is_valid a voter id from AC
+        voter_id = generate(federative_union ="AC")
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
        voter_id = generate(federative_union="AC")
```

File: tests/test_voter_id.py [hunk 29]
```diff
@@ -79,14 +80,33 @@ def test_calculate_vd1(self):
...
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if is_valid a voter id from foreigner
+        voter_id = generate()
+        self.assertIs(is_valid(voter_id), True)
+
+        # test if UF is not valid
+        voter_id = generate(federative_union ="XX")
```
Comment (camilamaia) [inline tests/test_voter_id.py]:
```suggestion
        voter_id = generate(federative_union="XX")
```

File: README.md [hunk 30]
```diff
@@ -897,6 +898,24 @@ False
 True
 ```
 
+### generate_voter_id
+
+Gera uma string de dígitos de Título de Eleitor válida aleatória a partir de um estado brasileiro informado.
+
+Retorna:
```
Comment (camilamaia) [inline README.md:908]:
```suggestion

Args:
  * federative_union (str): Unidade federativa para o título de eleitor que será gerado. O valor padrão "ZZ" é usado para títulos de eleitor emitidos para estrangeiros.

Retorna:
```

File: README_EN.md [hunk 31]
```diff
@@ -905,6 +906,26 @@ False
 True
 ```
 
+### generate_voter_id
+
+Generate a valid random Voter ID string of digits from an informed Brazilian federation union.
+
+Args:
+  * state(str): string representing the brazilian federation union.
```
Comment (camilamaia) [inline README_EN.md]:
```suggestion
  * federative_union(str): federative union for the voter id that will be generated. The default value "ZZ" is used for voter IDs issued to foreigners.
```

File: README_EN.md [hunk 32]
```diff
@@ -905,6 +906,26 @@ False
...
+Returns:
+  * str: A randomly generated valid voter ID.
+
+Example:
+
+```python
+>>> from brutils import generate_voter_id
+>>> generate_voter_id()
+'183475722801'
+>>> generate_voter_id(state="MG")
```
Comment (camilamaia) [inline README_EN.md]:
```suggestion
>>> generate_voter_id(federative_union ="MG")
```

File: README.md [hunk 33]
```diff
@@ -897,6 +898,24 @@ False
...
+  * str: Um Título de eleitor válido gerado aleatoriamente.
+
+Exemplo:
+
+```python
+>>> from brutils import generate_voter_id
+>>> generate_voter_id()
+'183475722801'
+>>> generate_voter_id(state="MG")
```
Comment (camilamaia) [inline README.md]:
```suggestion
>>> generate_voter_id(federative_union ="MG")
```

File: README.md [hunk 34]
```diff
@@ -897,6 +898,28 @@ False
 True
 ```
 
+### generate_voter_id
+
+Gera uma string de dígitos de Título de Eleitor válida aleatória a partir de um estado brasileiro informado.
+
```
Comment (camilamaia) [inline README.md:904]:
```suggestion
```

**Review by camilamaia — APPROVED** [review]
Oii @patricia-salles, mandou bem demais vendo que tinha faltado um edge case no cálculo do dv2, valeu!

A gente fez um pair aqui e arrumou os detalhes que faltavam, agora tá prontinho para dale!

Obrigada pela persistência, essa issue rendeu haha

**Comment by patricia-salles** [comment]
> Oii @patricia-salles, mandou bem demais vendo que tinha faltado um edge case no cálculo do dv2, valeu!
> 
> A gente fez um pair aqui e arrumou os detalhes que faltavam, agora tá prontinho para dale!
> 
> Obrigada pela persistência, essa issue rendeu haha

Que massa que consegui contribuir, fico feliz! Se tiver projetos futuros precisando de contribuição, posso tenta contribuir no meu tempo livre. Valeu :)
